cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+    {
+         Log(LOG_LEVEL_NOTICE, "Should install file type package: %s",
+             package_file_path);
+        res = PROMISE_RESULT_FAIL;
+    }
+    else
+    {
+        res = InstallPackage(policy_data->package_options,
+                             PACKAGE_TYPE_FILE, package_file_path,
+                             NULL, NULL, wrapper);
+        if (res == PROMISE_RESULT_CHANGE)
+        {
+            Log(LOG_LEVEL_DEBUG, "Validating package: %s", package_file_path);
+            return ValidateChangedPackage(policy_data, wrapper, info,
+                                          NEW_PACKAGE_ACTION_PRESENT);
+        }
+    }
+    return res;
+}
+
+
+Seq *GetVersionsFromUpdates(EvalContext *ctx, const PackageInfo *info,
+                            const PackageModuleWrapper *module_wrapper)
+{   
+    CF_DB *db_updates;
+    dbid db_id = dbid_packages_updates;
+    Seq *updates_list = NULL;
+    
+    /* Make sure cache is updated. */
+    if (!UpdateSinglePackageModuleCache(ctx, module_wrapper,
+                                        UPDATE_TYPE_UPDATES, false))
+    {
+        Log(LOG_LEVEL_INFO, "Can not update packages cache.");
+    }
+    
+    if (OpenSubDB(&db_updates, db_id, module_wrapper->package_module->name))
+    {
+        char package_key[strlen(info->name) + 4];
+
+        xsnprintf(package_key, sizeof(package_key),
+                "N<%s>", info->name);
+        
+        Log(LOG_LEVEL_DEBUG, "Looking for key in updates: %s", package_key);
+         
+        if (HasKeyDB(db_updates, package_key, sizeof(package_key)))
+        {
+            Log(LOG_LEVEL_DEBUG, "Found key in updates database");
+            
+            updates_list = SeqNew(3, FreePackageInfo);
+            size_t val_size =
+                    ValueSizeDB(db_updates, package_key, sizeof(package_key));
+            char buff[val_size + 1];
+            buff[val_size] = '\0';
+
+            ReadDB(db_updates, package_key, buff, val_size);
+            Seq* updates = SeqStringFromString(buff, '\n');
+            
+            for (int i = 0; i < SeqLength(updates); i++)
+            {
+                PackageInfo *package = calloc(1, sizeof(PackageInfo));
+
+                char *package_line = SeqAt(updates, i);
+                char version[strlen(package_line)];
+                char arch[strlen(package_line)];
+
+                if (sscanf(package_line, "V<%[^>]>A<%[^>]>", version, arch) == 2)
+                {
+                    package->version = SafeStringDuplicate(version);
+                    package->arch = SafeStringDuplicate(arch);
+                    SeqAppend(updates_list, package);
+                }
+                else
+                {
+                    Log(LOG_LEVEL_INFO, "Not able to parse available updates "
+                        "line: %s", package_line);
+                    /* Some error occurred while scanning package updates. */
+                    FreePackageInfo(package);
+                }
```

Comment:
I would put the `calloc()` and setting of `package->` fields into a separate `CreatePackageInfo(version, arch)` call, done inside the `sscanf`-success branch, saving the need to `FreePackageInfo()` in its `else` clause.